cmd/ocsp-responder/main.go

```diff
@@ -190,27 +232,202 @@ func (src *dbSource) Response(ctx context.Context, req *ocsp.Request) ([]byte, h
+			return nil, nil, err
+		}
+		src.log.Debugf("returning ocsp from primary source: %v", helper.PrettyResponse(primaryParsed))
+		src.ocspLookups.WithLabelValues("mysql_success").Inc()
+		src.sourceUsed.WithLabelValues("mysql").Inc()
+		return primaryResult.bytes, header, nil
+	case secondaryResult := <-secondaryChan:
+		// If secondary returns first, wait for primary to return for
+		// comparison.
+		var primaryResult lookupResponse
+
+		// Listen for cancellation or timeout waiting for primary result.
+		select {
+		case <-ctx.Done():
+			err := fmt.Errorf("looking up OCSP response for serial: %s err: %w", serialString, ctx.Err())
+			src.log.Debugf("%s", err)
+			src.ocspLookups.WithLabelValues("canceled").Inc()
+			return nil, nil, err
+		case primaryResult = <-primaryChan:
+		}
+
+		// Check for error returned from the mysql lookup, return on error.
+		if primaryResult.err != nil {
+			src.log.AuditErrf("Looking up OCSP response in DB: %s", err)
+			src.ocspLookups.WithLabelValues("mysql_failed").Inc()
+			src.sourceUsed.WithLabelValues("error_returned").Inc()
+			return nil, nil, primaryResult.err
+		}
+
+		// Parse the OCSP bytes returned from the primary source to check
+		// status, expiration and other fields.
+		primaryParsed, err := ocsp.ParseResponse(primaryResult.bytes, nil)
+		if err != nil {
+			src.log.AuditErrf("parsing OCSP response: %s", err)
+			src.ocspLookups.WithLabelValues("mysql_parse_error").Inc()
+			src.sourceUsed.WithLabelValues("error_returned").Inc()
+			return nil, nil, err
+		}
+
+		// Check for error returned from the redis lookup. If error return
+		// primary lookup result.
+		if secondaryResult.err != nil {
+			src.log.AuditErrf("Looking up OCSP response in redis: %s", err)
+			src.ocspLookups.WithLabelValues("redis_failed").Inc()
+			src.sourceUsed.WithLabelValues("mysql").Inc()
+			return primaryResult.bytes, header, nil
+		}
+
+		// Parse the OCSP bytes returned from the secondary source to
+		// compare to primary result.
+		secondaryParsed, err := ocsp.ParseResponse(secondaryResult.bytes, nil)
+		if err != nil {
+			src.log.Debugf("secondary OCSP lookup response error: %v", err)
+			src.ocspLookups.WithLabelValues("redis_parse_error").Inc()
+			src.sourceUsed.WithLabelValues("mysql").Inc()
+			return primaryResult.bytes, header, nil
+		}
+
+		// If the secondary response status doesn't match primary return
+		// primary response.
+		if primaryParsed.Status != secondaryParsed.Status {
+			src.ocspLookups.WithLabelValues("redis_mismatch").Inc()
+			src.sourceUsed.WithLabelValues("mysql").Inc()
+			return primaryResult.bytes, header, nil
+		}
+
+		// The secondary response has passed checks, return it.
+		src.ocspLookups.WithLabelValues("redis_success").Inc()
+		src.sourceUsed.WithLabelValues("redis").Inc()
+		return secondaryResult.bytes, header, nil
 	}
-	return certStatus.OCSPResponse, header, nil
 }
 
+// ocspLookup has a getResponse method that knows how to retrieve and OCSP
```

Comment:
```suggestion
// ocspLookup has a getResponse method that knows how to retrieve an OCSP
```

Reply:
I got rid of main.Responder to avoid confusion, and fixed the grammar error. Thanks!